File: src/main/java/komu/blunt/types/checker/TypeUtils.java
package komu.blunt.types.checker;

import komu.blunt.types.TypeVariable;
import komu.blunt.types.Types;

import java.util.*;

public final class TypeUtils {

    public static void addTypeVariables(Set<TypeVariable> variables, Iterable<? extends Types<?>> ts) {
        for (Types<?> t : ts)
            t.addTypeVariables(variables);
    }
    
    public static Set<TypeVariable> getTypeVariables(Types<?> t) {
        LinkedHashSet<TypeVariable> types = new LinkedHashSet<TypeVariable>();
        t.addTypeVariables(types);
        return types;
    }

    public static <T extends Types<T>> List<T> applySubstitution(Substitution substitution, Collection<? extends T> ts) {
        List<T> result = new ArrayList<T>(ts.size());
        
        for (T t : ts)
            result.add(t.apply(substitution));

        return result;
    }
}

File: src/main/kotlin/komu/blunt/types/TypeUtils.kt
package komu.blunt.types

import komu.blunt.types.checker.Substitution

object TypeUtils {

    fun addTypeVariables<T : Types<T>>(variables: MutableSet<TypeVariable>, ts: Collection<T>) {
        for (val t in ts)
            t.addTypeVariables(variables)
    }

    fun <T : Types<T>> applySubstitution(substitution: Substitution, ts: Collection<T>): List<T> =
        ts.map { it.apply(substitution) }
}

